org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/batch/ClasspathJsr199.java

```diff
@@ -156,12 +160,29 @@ public char[][][] findTypeNames(String aQualifiedPackageName, String moduleName)
 	public void initialize() throws IOException {
 		if (this.jrt != null) {
 			this.jrt.initialize();
+		} else if (this.location.isModuleOrientedLocation()) {
+			Iterable<Set<Location>> locationsForModules = this.fileManager.listLocationsForModules(this.location);
+			for (Set<Location> locs: locationsForModules) {
+				for (Location loc : locs) {
+					if (loc instanceof LocationWrapper wrapper) {
+						for (Path locPath : wrapper.getPaths()) {
+							File file = locPath.toFile();
+							IModule mod = ModuleFinder.scanForModule(this, file, null, true, null);
+							if (mod != null) {
+								return;
```

Comment:
Several locations now assume that each ClasspathJsr199 is either JRT or a single module classpath location. See in particular new uses of the inherited field `module`. At a second look this is probably wrong, and each ClasspathJsr199 must be able to handle multiple modules.

Is this initialization the place that should actually create a `Map<Location,IModule>`? I'm a bit confused by the multi-level multiplicities: an Iterable of Sets of Locations where each Location has multiple Paths. At which of these levels can we assume a unique module? Is a LocationWrapper the right level? So if a module is in a jar, can we assume that it is represented by one LocationWrapper containing exactly one path?

Reply:
Obsoleted by creating one `ClasspathJsr199` for each module.